specs/approved/power-control-intermediary.rst

```diff
@@ -0,0 +1,741 @@
+users to perform this sort of action, the absence of this functionality is a
+potential roadblock for improving multi-tenant support. [0]_ [1]_
+
+Since giving a node lessee direct BMC access is very unwise, we instead intend
+to emulate a standards-based interface (preferably Redfish, although IPMI is
+another valid option) directly within Ironic for use by node lessees. This
+way, Ironic can handle authenticating this request directly and provisioning
+workflows that require this type of API access can function as expected.
+
+
+Problem description
+===================
+
+As it stands, the only way an unprivileged user, such as a lessee, can access
+power management APIs would be directly through the BMC with the node's power
+credentials. But since these credentials wouldn't necessarily change, the
+lessee could re-use them to change power settings after they were supposed to
+have lost access.
+
+In a situation like this, the lessee would be forced to use Ironic's API and
+provisioning tools, possibly requiring existing workflows to be completely
+refactored. There are many potential cases in which this would be very
+problematic to the end-user, and the alternative solution of providing the
+lessee with power credentials would be a huge security risk unacceptable to
+most administrators.
+
+
+Proposed change
+===============
+
+We will create a new REST API endpoint at ``{IRONIC URI}/redfish/v1`` that
+will act like a legitimate v1 Redfish endpoint to the end user. The
+functionality of this interface will provide strictly the essentials for
+Redfish-based provisioning tools to access power controls, as implementation
+of a full Redfish proxy would be beyond the scope of what we intend to do.
+(see `REST API Impact`_ below for more details). Minimally, it will allow for
+a user to identify themselves to Ironic and get/set the power state of any
+nodes that they have access to.
+
+These pseudo-Redfish API calls will be be translated into corresponding Ironic
+or Keystone API calls, and will allow for proper authentication as well as
+access to power controls. Provisioning capabilities will be available through
+Redfish's ``ComputerSystem`` interface and will provide equivalent
+functionality to the following Ironic CLI commands:
+
+* ``baremetal node list``
+* ``baremetal node show``
+* ``baremetal node power on``
+* ``baremetal node power off``
+* ``baremetal ndoe reboot``
+
+The workflow for authentication will be as follows:
+
+* User creates an application credential [2]_ for the project that has access
+  to the node they wish to access.
+* User uses the Redfish Sessions API to authenticate, specifying the
+  application credential UUID as the username, and the credential's secret
+  as the password.
+* User receives a token that grants them access to provisioning actions that
+  require authentication
+* Session is deleted; any further provisioning actions will require
```

Comment:
Might make sense to add a step before "session is deleted" where users actually use the token and the Ironic redfish proxy

Reply:
that's _maybe_ an important step... has been added :)